core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1941,11 +1941,21 @@ class KafkaApis(val requestChannel: RequestChannel,
     // with client authentication which is performed at an earlier stage of the connection where the
     // ApiVersionRequest is not available.
     val apiVersionRequest = request.body[ApiVersionsRequest]
+    val expectedClientSoftwareNames = config.expectedClientSoftwareNames
+    var hasNonXinfraClient = false
+    if (expectedClientSoftwareNames != null) {
+      expectedClientSoftwareNames.forEach { clientName =>
+        if (!clientName.toLowerCase.contains("xinfra")) {
+          hasNonXinfraClient = true;
```

Comment:
should these information also be cached? Since handleAPIVersionRequest is only invoked once during clients init connection; all future produce/consume requests will not include this clientName, only client id from the context; I assume you'll need observer to observe produce/consume requests?

Reply:
this information is cached within observer based on the clientId. For now we're only flagging whether it is a xinfra client or not but future iterations we plan on adding a string that has all client names that have accessed this topic (this is in the design doc but we can discuss more if you want). Are you suggesting another layer of caching inside kafkaApis so if a clientId which already accessed this topic attempts to access again we won't check it and won't store it in observer?

Reply:
if this info is cached within observer based on clientId then it's fine, just want to make sure in other places in KafkaApis like handleProduce/handleConsume, when we call observer.observe() it can distinguish LKC clients vs XC clients

Reply:
yes it will be able to distinguish that